File: src/libiec61850/dotnet/core/2.0/iec61850_client/TagsCreation.cs
/* 
 * This software implements a IEC61850 driver for JSON SCADA.
 * Copyright - 2020-2024 - Ricardo Lastra Olsen
 * 
 * Requires libiec61850 from MZ Automation.
 * 
 * This program is free software: you can redistribute it and/or modify  
 * it under the terms of the GNU General Public License as published by  
 * the Free Software Foundation, version 3.
 *
 * This program is distributed in the hope that it will be useful, but 
 * WITHOUT ANY WARRANTY; without even the implied warranty of 
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU 
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License 
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

using System;
using MongoDB.Bson;
using MongoDB.Bson.Serialization;
using MongoDB.Bson.Serialization.Attributes;

namespace IEC61850_Client
{
    partial class MainClass
    {
        [BsonIgnoreExtraElements]
        public class rtDataId
        {
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(0.0)]
            public BsonDouble _id;
        }
        [BsonIgnoreExtraElements]
        public class rtData
        {
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(0.0)]
            public BsonDouble _id;
            [BsonDefaultValue(false)]
            public BsonBoolean alarmDisabled { get; set; }
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(0.0)]
            public BsonDouble alarmState;
            [BsonDefaultValue(false)]
            public BsonBoolean alarmed { get; set; }
            [BsonDefaultValue(false)]
            public BsonBoolean alerted { get; set; }
            [BsonDefaultValue("")]
            public BsonString alertState { get; set; }
            [BsonDefaultValue("")]
            public BsonString annotation { get; set; }
            [BsonDefaultValue(false)]
            public BsonBoolean commandBlocked { get; set; }
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(0.0)]
            public BsonDouble commandOfSupervised;
            [BsonDefaultValue("")]
            public BsonString commissioningRemarks { get; set; }
            [BsonDefaultValue("")]
            public BsonString description { get; set; }
            [BsonDefaultValue("")]
            public BsonString eventTextFalse { get; set; }
            [BsonDefaultValue("")]
            public BsonString eventTextTrue { get; set; }
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(0.0)]
            public BsonDouble formula;
            [BsonDefaultValue(false)]
            public BsonBoolean frozen { get; set; }
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(0.0)]
            public BsonDouble frozenDetectTimeout;
            [BsonDefaultValue("")]
            public BsonString group1 { get; set; }
            [BsonDefaultValue("")]
            public BsonString group2 { get; set; }
            [BsonDefaultValue("")]
            public BsonString group3 { get; set; }
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(Double.MaxValue)]
            public BsonDouble hiLimit;
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(Double.MaxValue)]
            public BsonDouble hihiLimit;
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(Double.MaxValue)]
            public BsonDouble hihihiLimit;
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(0.0)]
            public BsonDouble historianDeadBand;
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(0.0)]
            public BsonDouble historianPeriod;
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(0.0)]
            public BsonDouble hysteresis;
            [BsonDefaultValue(true)]
            public BsonBoolean invalid { get; set; }
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(60000.0)]
            public BsonDouble invalidDetectTimeout;
            [BsonDefaultValue(false)]
            public BsonBoolean isEvent { get; set; }
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(1.0)]
            public BsonDouble kconv1;
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(0.0)]
            public BsonDouble kconv2;
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(-Double.MaxValue)]
            public BsonDouble loLimit;
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(-Double.MaxValue)]
            public BsonDouble loloLimit;
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(-Double.MaxValue)]
            public BsonDouble lololoLimit;
            [BsonDefaultValue(null)]
            public BsonValue location;
            [BsonDefaultValue("")]
            public BsonString notes { get; set; }
            [BsonDefaultValue("supervised")]
            public BsonString origin { get; set; }
            [BsonDefaultValue(false)]
            public BsonBoolean overflow { get; set; }
            [BsonDefaultValue(null)]
            public BsonValue parcels;
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(0.0)]
            public BsonDouble priority;
            [BsonDefaultValue(null)]
            public BsonValue protocolDestinations;
            [BsonDefaultValue("")]
            public BsonString protocolSourceASDU { get; set; }
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(0.0)]
            public BsonDouble protocolSourceCommandDuration;
            [BsonDefaultValue(false)]
            public BsonBoolean protocolSourceCommandUseSBO { get; set; }
            [BsonDefaultValue("")]
            public BsonString protocolSourceCommonAddress { get; set; }
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(0.0)]
            public BsonDouble protocolSourceConnectionNumber;
            [BsonDefaultValue("")]
            public BsonString protocolSourceObjectAddress { get; set; }

            [BsonDefaultValue(null)]
            public BsonValue sourceDataUpdate { get; set; }
            [BsonDefaultValue("")]
            public BsonString stateTextFalse { get; set; }
            [BsonDefaultValue("")]
            public BsonString stateTextTrue { get; set; }
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(0.0)]
            public BsonDouble supervisedOfCommand;
            [BsonDefaultValue("")]
            public BsonString tag { get; set; }
            [BsonDefaultValue(null)]
            public BsonValue timeTag { get; set; }
            [BsonDefaultValue(null)]
            public BsonValue timeTagAlarm { get; set; }
            [BsonDefaultValue(null)]
            public BsonValue timeTagAtSource { get; set; }
            [BsonDefaultValue(false)]
            public BsonBoolean timeTagAtSourceOk { get; set; }
            [BsonDefaultValue(false)]
            public BsonBoolean transient { get; set; }
            [BsonDefaultValue("digital")]
            public BsonString type { get; set; }
            [BsonDefaultValue("")]
            public BsonString ungroupedDescription { get; set; }
            [BsonDefaultValue("")]
            public BsonString unit { get; set; }
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(0.0)]
            public BsonDouble updatesCnt;
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(0.0)]
            public BsonDouble valueDefault;
            [BsonDefaultValue("")]
            public BsonString valueString { get; set; }
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(0.0)]
            public BsonDouble value;
            [BsonSerializer(typeof(BsonDoubleSerializer)), BsonDefaultValue(0.0)]
            public BsonDouble zeroDeadband;
        }
        public static rtData newRealtimeDoc(IECValue iv, double _id)
        {
            var group1 = "IEC61850";
            if (iv.asdu.ToLower() == "boolean" || iv.isDigital)
                return new rtData()
                {
                    _id = _id,
                    protocolSourceASDU = iv.asdu,
                    protocolSourceCommonAddress = iv.common_address.ToUpper(),
                    protocolSourceConnectionNumber = iv.conn_number,
                    protocolSourceObjectAddress = iv.address,
                    protocolSourceCommandUseSBO = false,
                    protocolSourceCommandDuration = 0.0,
                    alarmState = 2.0,
                    description = group1 + "~" + iv.conn_name + "~" + iv.display_name,
                    ungroupedDescription = iv.display_name,
                    group1 = group1,
                    group2 = iv.conn_name,
                    group3 = iv.common_address,
                    stateTextFalse = "FALSE",
                    stateTextTrue = "TRUE",
                    eventTextFalse = "FALSE",
                    eventTextTrue = "TRUE",
                    origin = "supervised",
                    tag = TagFromParameters(iv),
                    type = "digital",
                    value = iv.value,
                    valueString = "????",
                    alarmDisabled = false,
                    alerted = false,
                    alarmed = false,
                    alertState = "",
                    annotation = "",
                    commandBlocked = false,
                    commandOfSupervised = 0.0,
                    commissioningRemarks = "",
                    formula = 0.0,
                    frozen = false,
                    frozenDetectTimeout = 0.0,
                    hiLimit = Double.MaxValue,
                    hihiLimit = Double.MaxValue,
                    hihihiLimit = Double.MaxValue,
                    historianDeadBand = 0.0,
                    historianPeriod = 0.0,
                    hysteresis = 0.0,
                    invalid = true,
                    invalidDetectTimeout = 60000,
                    isEvent = false,
                    kconv1 = 1.0,
                    kconv2 = 0.0,
                    location = BsonNull.Value,
                    loLimit = -Double.MaxValue,
                    loloLimit = -Double.MaxValue,
                    lololoLimit = -Double.MaxValue,
                    notes = "",
                    overflow = false,
                    parcels = BsonNull.Value,
                    priority = 0.0,
                    protocolDestinations = BsonNull.Value,
                    sourceDataUpdate = BsonNull.Value,
                    supervisedOfCommand = 0.0,
                    timeTag = BsonNull.Value,
                    timeTagAlarm = BsonNull.Value,
                    timeTagAtSource = BsonNull.Value,
                    timeTagAtSourceOk = false,
                    transient = false,
                    unit = "",
                    updatesCnt = 0,
                    valueDefault = 0.0,
                    zeroDeadband = 0.0
                };
            else
            if (iv.asdu.ToLower() == "string" || iv.asdu.ToLower() == "extensionobject")
                return new rtData()
                {
                    _id = _id,
                    protocolSourceASDU = iv.asdu,
                    protocolSourceCommonAddress = iv.common_address.ToUpper(),
                    protocolSourceConnectionNumber = iv.conn_number,
                    protocolSourceObjectAddress = iv.address,
                    protocolSourceCommandUseSBO = false,
                    protocolSourceCommandDuration = 0.0,
                    alarmState = -1.0,
                    description = group1 + "~" + iv.conn_name + "~" + iv.display_name,
                    ungroupedDescription = iv.display_name,
                    group1 = group1,
                    group2 = iv.conn_name,
                    group3 = iv.common_address,
                    stateTextFalse = "",
                    stateTextTrue = "",
                    eventTextFalse = "",
                    eventTextTrue = "",
                    origin = "supervised",
                    tag = TagFromParameters(iv),
                    type = "string",
                    value = 0.0,
                    valueString = iv.valueString,

                    alarmDisabled = false,
                    alerted = false,
                    alarmed = false,
                    alertState = "",
                    annotation = "",
                    commandBlocked = false,
                    commandOfSupervised = 0.0,
                    commissioningRemarks = "",
                    formula = 0.0,
                    frozen = false,
                    frozenDetectTimeout = 0.0,
                    hiLimit = Double.MaxValue,
                    hihiLimit = Double.MaxValue,
                    hihihiLimit = Double.MaxValue,
                    historianDeadBand = 0.0,
                    historianPeriod = 0.0,
                    hysteresis = 0.0,
                    invalid = true,
                    invalidDetectTimeout = 60000,
                    isEvent = false,
                    kconv1 = 1.0,
                    kconv2 = 0.0,
                    location = BsonNull.Value,
                    loLimit = -Double.MaxValue,
                    loloLimit = -Double.MaxValue,
                    lololoLimit = -Double.MaxValue,
                    notes = "",
                    overflow = false,
                    parcels = BsonNull.Value,
                    priority = 0.0,
                    protocolDestinations = BsonNull.Value,
                    sourceDataUpdate = BsonNull.Value,
                    supervisedOfCommand = 0.0,
                    timeTag = BsonNull.Value,
                    timeTagAlarm = BsonNull.Value,
                    timeTagAtSource = BsonNull.Value,
                    timeTagAtSourceOk = false,
                    transient = false,
                    unit = "",
                    updatesCnt = 0,
                    valueDefault = 0.0,
                    zeroDeadband = 0.0,
                };

            return new rtData()
            {
                _id = _id,
                protocolSourceASDU = iv.asdu,
                protocolSourceCommonAddress = iv.common_address.ToUpper(),
                protocolSourceConnectionNumber = iv.conn_number,
                protocolSourceObjectAddress = iv.address,
                protocolSourceCommandUseSBO = false,
                protocolSourceCommandDuration = 0.0,
                alarmState = -1.0,
                description = group1 + "~" + iv.conn_name + "~" + iv.display_name,
                ungroupedDescription = iv.display_name,
                group1 = group1,
                group2 = iv.conn_name,
                group3 = iv.common_address,
                stateTextFalse = "",
                stateTextTrue = "",
                eventTextFalse = "",
                eventTextTrue = "",
                origin = "supervised",
                tag = TagFromParameters(iv),
                type = "analog",
                value = iv.value,
                valueString = "????",

                alarmDisabled = false,
                alerted = false,
                alarmed = false,
                alertState = "",
                annotation = "",
                commandBlocked = false,
                commandOfSupervised = 0.0,
                commissioningRemarks = "",
                formula = 0.0,
                frozen = false,
                frozenDetectTimeout = 0.0,
                hiLimit = Double.MaxValue,
                hihiLimit = Double.MaxValue,
                hihihiLimit = Double.MaxValue,
                historianDeadBand = 0.0,
                historianPeriod = 0.0,
                hysteresis = 0.0,
                invalid = true,
                invalidDetectTimeout = 60000,
                isEvent = false,
                kconv1 = 1.0,
                kconv2 = 0.0,
                location = BsonNull.Value,
                loLimit = -Double.MaxValue,
                loloLimit = -Double.MaxValue,
                lololoLimit = -Double.MaxValue,
                notes = "",
                overflow = false,
                parcels = BsonNull.Value,
                priority = 0.0,
                protocolDestinations = BsonNull.Value,
                sourceDataUpdate = BsonNull.Value,
                supervisedOfCommand = 0.0,
                timeTag = BsonNull.Value,
                timeTagAlarm = BsonNull.Value,
                timeTagAtSource = BsonNull.Value,
                timeTagAtSourceOk = false,
                transient = false,
                unit = "",
                updatesCnt = 0,
                valueDefault = 0.0,
                zeroDeadband = 0.0
            };
        }
        static string TagFromParameters(IECValue iv)
        {
            return "IEC61850;" + iv.conn_name + ";" + iv.address + "[" + iv.common_address + "]";
        }
    }
}

File: src/iec61850/iec61850_client/tags_creation.go
/*
 * IEC 61850 MMS Client driver for {json:scada}, in Go.
 * {json:scada} - Copyright (c) 2020-2026 - Ricardo L. Olsen
 * This file is part of the JSON-SCADA distribution (https://github.com/riclolsen/json-scada).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

// Automatic tag creation. Port of TagsCreation.cs: the documents inserted
// here must match what the C# driver inserts, field for field, so that a
// database seeded by one driver works with the other.

package main

import (
	"maps"
	"strings"

	"github.com/riclolsen/json-scada/src/go-common/jstags"
	"go.mongodb.org/mongo-driver/v2/bson"
)

// TagFromParameters builds the tag name of an automatically created point.
func TagFromParameters(iv IECValue) string {
	return "IEC61850;" + iv.ConnName + ";" + iv.Address + "[" + iv.CommonAddress + "]"
}

// CommandTag describes a controllable object found while browsing, waiting
// for its tag to be created.
type CommandTag struct {
	ConnNumber int
	ConnName   string
	Ref        string // IEC 61850 object reference of the control object
	IsDigital  bool
	UseSBO     bool
	Asdu       string // MMS type of the control value
	Attempts   int    // times the supervised twin was looked for
}

// Tag is the name of the command tag.
func (c CommandTag) Tag() string {
	return "IEC61850;" + c.ConnName + ";" + c.Ref + "[CO]"
}

// newCommandDoc builds the realtimeData document of a command point.
// supervisedID is the key of the point where the command's effect shows,
// zero when the device exposes no status for the controllable object.
func newCommandDoc(ct CommandTag, id, supervisedID float64) bson.M {
	const group1 = "IEC61850"

	doc := jstags.BaseDoc()
	maps.Copy(doc, bson.M{
		"_id":                            id,
		"protocolSourceASDU":             ct.Asdu,
		"protocolSourceCommonAddress":    "CO",
		"protocolSourceConnectionNumber": float64(ct.ConnNumber),
		"protocolSourceObjectAddress":    ct.Ref,
		"protocolSourceCommandUseSBO":    ct.UseSBO,
		"protocolSourceCommandDuration":  0.0,
		"description":                    group1 + "~" + ct.ConnName + "~" + ct.Ref + " command",
		"ungroupedDescription":           ct.Ref + " command",
		"group1":                         group1,
		"group2":                         ct.ConnName,
		"group3":                         "CO",
		"origin":                         "command",
		"tag":                            ct.Tag(),
		// The two ends of the pair: this command acts on that supervised
		// point, which is where its feedback appears.
		"supervisedOfCommand":  supervisedID,
		"commandOfSupervised":  0.0,
		"invalid":              false,
		"invalidDetectTimeout": 0.0,
		"protocolDestinations": nil,
		"value":                0.0,
		"valueString":          "",
	})

	if ct.IsDigital {
		doc["type"] = "digital"
		doc["alarmState"] = -1.0
		doc["stateTextFalse"] = "FALSE"
		doc["stateTextTrue"] = "TRUE"
		doc["eventTextFalse"] = "FALSE"
		doc["eventTextTrue"] = "TRUE"
	} else {
		doc["type"] = "analog"
		doc["alarmState"] = -1.0
		doc["stateTextFalse"] = ""
		doc["stateTextTrue"] = ""
		doc["eventTextFalse"] = ""
		doc["eventTextTrue"] = ""
	}
	return doc
}

// newRealtimeDoc builds the realtimeData document for a discovered point.
func newRealtimeDoc(iv IECValue, id float64) bson.M {
	const group1 = "IEC61850"

	doc := jstags.BaseDoc()
	maps.Copy(doc, bson.M{
		"_id":                            id,
		"protocolSourceASDU":             iv.Asdu,
		"protocolSourceCommonAddress":    strings.ToUpper(iv.CommonAddress),
		"protocolSourceConnectionNumber": float64(iv.ConnNumber),
		"protocolSourceObjectAddress":    iv.Address,
		"protocolSourceCommandUseSBO":    false,
		"protocolSourceCommandDuration":  0.0,
		"description":                    group1 + "~" + iv.ConnName + "~" + iv.DisplayName,
		"ungroupedDescription":           iv.DisplayName,
		"group1":                         group1,
		"group2":                         iv.ConnName,
		"group3":                         iv.CommonAddress,
		"origin":                         "supervised",
		"tag":                            TagFromParameters(iv),
		"commandOfSupervised":            0.0,
		"invalid":                        true,
		"invalidDetectTimeout":           60000.0,
		"protocolDestinations":           nil,
		"supervisedOfCommand":            0.0,
	})

	switch {
	case strings.EqualFold(iv.Asdu, "boolean") || iv.IsDigital:
		doc["alarmState"] = 2.0
		doc["stateTextFalse"] = "FALSE"
		doc["stateTextTrue"] = "TRUE"
		doc["eventTextFalse"] = "FALSE"
		doc["eventTextTrue"] = "TRUE"
		doc["type"] = "digital"
		doc["value"] = iv.Value
		doc["valueString"] = "????"
	case strings.EqualFold(iv.Asdu, "string") || strings.EqualFold(iv.Asdu, "extensionobject"):
		doc["alarmState"] = -1.0
		doc["stateTextFalse"] = ""
		doc["stateTextTrue"] = ""
		doc["eventTextFalse"] = ""
		doc["eventTextTrue"] = ""
		doc["type"] = "string"
		doc["value"] = 0.0
		doc["valueString"] = iv.ValueString
	default:
		doc["alarmState"] = -1.0
		doc["stateTextFalse"] = ""
		doc["stateTextTrue"] = ""
		doc["eventTextFalse"] = ""
		doc["eventTextTrue"] = ""
		doc["type"] = "analog"
		doc["value"] = iv.Value
		doc["valueString"] = "????"
	}

	return doc
}
